### DISCUSSION TITLE
[Proposal]: Explicit Interface Implementation for property setter

### DISCUSSION THREAD

**Discussion by znakeeye** [dotnet/csharplang#8829] [Language Ideas]
# Explicit Interface Implementation for property setter

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

We can easily hide a method using [Explicit Interface Implementation](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/interfaces/explicit-interface-implementation). It makes sense to allow a similar syntax for an interfaced property-setter. E.g.:

```c#
IDisposable.Dispose() {} // OK
public int MyProp { get; IMyInterface.MyProp.set } // Suggestion
```

## Motivation
[motivation]: #motivation
There are situations where you want to be able to set a property value, but for the general implementation you want to hide this capability. Similar to the `IDisposable` pattern where you might want only dispose-aware code to be able to call `Dispose()`. The naïve solution, is to define an interface like this and then implement it with a no-op setter. I'm including this since I have seen it in many projects (apparently not all developers are aware of the *explicit interface implementation* feature).

```c#
public interface IMyInterface
{
    int MyProp { get; set; }
}

public class NaiveClass : IMyInterface
{
    private int myProp;
    public int MyProp
    {
        get => myProp;
        set => throw new NotImplementedException();
    }
}
```

This naïve solution will allow access to the setter, but will cause a runtime failure if called. An improved version is shown below:

```c#
public class BetterClass : IMyInterface
{
    public int Property { get; private set; }
    int IMyInterface.Property
    {
        get => Property;
        set => Property = value;
    }
}
```

Too much typing! With my suggestion, you can define the property in a one-liner - just like those other nice property declarations!

## Detailed design
[design]: #detailed-design

## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

I can think of a few syntax alternatives.

```c#
public int MyProp { get; IMyInterface.MyProp.set }
// or
public int MyProp { get; IMyInterface.set }
// or
public int MyProp { get; explicit set; }
```

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by YairHalberstadt** [comment]
Given how rare this need is, the current code seems like it's fine:

```csharp
public class BetterClass : IMyInterface
{
    public int Property { get; private set; }
    int IMyInterface.Property
    {
        get => Property;
        set => Property = value;
    }
}
```

**Comment by PathogenDavid** [comment]
And if it's really common in your problem domain, a source generator could make quick work of the boilerplate.

**Comment by msedi** [comment]
It seems to me that you would need some sort of a read-only implementation (I cannot find the issue/discussion around this). I solved this by having read-only interfaces, regular interfaces and then some implementation (see below). But you are right. It's way too much typing for such a simple requirement.

```cs
    public interface IReadOnlyVolume
    {
        int SizeX { get; }
        int SizeY { get; }
        int SizeZ { get; }
    }

    public interface IVolume : IReadOnlyVolume
    {
        new int SizeX { get; set; }
        new int SizeY { get; set; }
        new int SizeZ { get; set; }
    }

    public class Volume : IVolume
    {
        public int SizeX {get;set;}
        public int SizeY {get;set;}
        public int SizeZ { get; set; }

        int IReadOnlyVolume.SizeX => SizeX;
        int IReadOnlyVolume.SizeY => SizeY;
        int IReadOnlyVolume.SizeZ => SizeZ;
    }

```

**Comment by znakeeye** [comment]
> I solved this by having read-only interfaces, regular interfaces and then some implementation (see below).

Exactly. This is what I wanted to implement, but with code bloat.

**Comment by msedi** [comment]
@znakeeye: There is a discussion around deep immutability: #4015. Although I understand the problems that are discussed there I truly wish that something like this exists.
